# Data Wrangling for WeRateDogs twitter handle

WeRateDogs is a Twitter handle that rates dogs on a scale of 10 and posts it along with a picture or video submitted to the account by the owner. Rating often go above 10 and this has definitely made this rating system unique. The twitter handle has above 6.9 million followers and an online shopping store for dogs and dog owners that relies on marketing through this handle. 

In order to analyze the twitter feeds of for this account it is important to wrangle a dataset for it. In this notebook, you will be able to see the data wrangling efforts made before analyzing the twitter feeds of WeRateDogs.

**Data wrangling process**
The data wrangling process for this projectt will be divided into three major components:

1. Gather
    - Involves gathering different datasets needed to do the analysis. These datasets will be stored in different dataframes.
2. Assess
    - The gathered data will be assessed for cleaning and tidiness issues and all problems will be documented for further cleaning
3. Clean
    - Retify cleaning and tidiness issues that are documented in the assessment step

## Project 

In [1]:
#imports
import numpy as np
import pandas as pd
import requests
import io
import tweepy
import json
import os
import time
import re
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
sns.set_style('darkgrid' )

## Gather

**Reading in the twitter archive**

In [2]:
df_1=pd.read_csv("twitter-archive-enhanced.csv")

**Retrieving the image prediction data from cloud**

In [3]:
# Used read_csv instead of the requests library since it seemed more efficient and could be done using less code
# the udacity mentor said it should be alright
url = "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
df_2= pd.read_csv(url, sep = "\t")


**Accessing Twitter API to get additional data**

Authorization for access

In [4]:
# the information is stored in txt file line by line
with open("tokens.txt") as token:
    consumer_key = token.readline().strip()
    consumer_secret = token.readline().strip()
    access_token = token.readline().strip()
    access_secret = token.readline().strip()

In [5]:
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

Saving twitter API data in a .txt file for use, to avoid repeatedly querying the API since it takes long
- can also be done using pickling 

In [6]:
tweet_id_list=df_1['tweet_id'].tolist()

#appends json by line but also adds \n at the end 
start = time.time()
error_tweet_list=[]
success_tweets= []
with open('tweet_json.txt', 'a') as tweet_json:
    for count,tweet in enumerate(tweet_id_list):
        try:
            status_obj = api.get_status(tweet, tweet_mode='extended')
            json_str = json.dumps(status_obj._json)
            tweet_json.write(json_str + "\n")
            success_tweets.append(tweet)
            print (str(count)+". "+str(tweet)+": "+"Success") 
        except: 
            error_tweet_list.append(tweet)
            print (str(count)+". "+str(tweet)+": "+"FAIL! Tweet ID not found.")  
end = time.time()
print(str(end - start)+"seconds to complete")

0. 892420643555336193: Success
1. 892177421306343426: Success
2. 891815181378084864: Success
3. 891689557279858688: Success
4. 891327558926688256: Success
5. 891087950875897856: Success
6. 890971913173991426: Success
7. 890729181411237888: Success
8. 890609185150312448: Success
9. 890240255349198849: Success
10. 890006608113172480: Success
11. 889880896479866881: Success
12. 889665388333682689: Success
13. 889638837579907072: Success
14. 889531135344209921: Success
15. 889278841981685760: Success
16. 888917238123831296: Success
17. 888804989199671297: Success
18. 888554962724278272: Success
19. 888202515573088257: FAIL! Tweet ID not found.
20. 888078434458587136: Success
21. 887705289381826560: Success
22. 887517139158093824: Success
23. 887473957103951883: Success
24. 887343217045368832: Success
25. 887101392804085760: Success
26. 886983233522544640: Success
27. 886736880519319552: Success
28. 886680336477933568: Success
29. 886366144734445568: Success
30. 886267009285017600: Success


249. 845306882940190720: Success
250. 845098359547420673: Success
251. 844979544864018432: Success
252. 844973813909606400: Success
253. 844704788403113984: Success
254. 844580511645339650: Success
255. 844223788422217728: Success
256. 843981021012017153: Success
257. 843856843873095681: Success
258. 843604394117681152: Success
259. 843235543001513987: Success
260. 842892208864923648: FAIL! Tweet ID not found.
261. 842846295480000512: Success
262. 842765311967449089: Success
263. 842535590457499648: Success
264. 842163532590374912: Success
265. 842115215311396866: Success
266. 841833993020538882: Success
267. 841680585030541313: Success
268. 841439858740625411: Success
269. 841320156043304961: Success
270. 841314665196081154: Success
271. 841077006473256960: Success
272. 840761248237133825: Success
273. 840728873075638272: Success
274. 840698636975636481: Success
275. 840696689258311684: Success
276. 840632337062862849: Success
277. 840370681858686976: Success
278. 840268004936019968: 

496. 813157409116065792: Success
497. 813142292504645637: Success
498. 813130366689148928: Success
499. 813127251579564032: Success
500. 813112105746448384: Success
501. 813096984823349248: Success
502. 813081950185472002: Success
503. 813066809284972545: Success
504. 813051746834595840: Success
505. 812781120811126785: Success
506. 812747805718642688: Success
507. 812709060537683968: Success
508. 812503143955202048: Success
509. 812466873996607488: Success
510. 812372279581671427: Success
511. 811985624773361665: Success
512. 811744202451197953: Success
513. 811647686436880384: Success
514. 811627233043480576: Success
515. 811386762094317568: Success
516. 810984652412424192: Success
517. 810896069567610880: Success
518. 810657578271330305: Success
519. 810284430598270976: Success
520. 810254108431155201: Success
521. 809920764300447744: Success
522. 809808892968534016: Success
523. 809448704142938112: Success
524. 809220051211603969: Success
525. 809084759137812480: Success
526. 80883

744. 780192070812196864: Success
745. 780092040432480260: Success
746. 780074436359819264: Success
747. 779834332596887552: Success
748. 779377524342161408: Success
749. 779124354206535695: Success
750. 779123168116150273: Success
751. 779056095788752897: Success
752. 778990705243029504: Success
753. 778774459159379968: Success
754. 778764940568104960: Success
755. 778748913645780993: Success
756. 778650543019483137: Success
757. 778624900596654080: Success
758. 778408200802557953: Success
759. 778396591732486144: Success
760. 778383385161035776: Success
761. 778286810187399168: Success
762. 778039087836069888: Success
763. 778027034220126208: Success
764. 777953400541634568: Success
765. 777885040357281792: Success
766. 777684233540206592: Success
767. 777641927919427584: Success
768. 777621514455814149: Success
769. 777189768882946048: Success
770. 776819012571455488: Success
771. 776813020089548800: Success
772. 776477788987613185: Success
773. 776249906839351296: Success
774. 77621

992. 748692773788876800: Success
993. 748575535303884801: Success
994. 748568946752774144: Success
995. 748346686624440324: Success
996. 748337862848962560: Success
997. 748324050481647620: Success
998. 748307329658011649: Success
999. 748220828303695873: Success
1000. 747963614829678593: Success
1001. 747933425676525569: Success
1002. 747885874273214464: Success
1003. 747844099428986880: Success
1004. 747816857231626240: Success
1005. 747651430853525504: Success
1006. 747648653817413632: Success
1007. 747600769478692864: Success
1008. 747594051852075008: Success
1009. 747512671126323200: Success
1010. 747461612269887489: Success
1011. 747439450712596480: Success
1012. 747242308580548608: Success
1013. 747219827526344708: Success
1014. 747204161125646336: Success
1015. 747103485104099331: Success
1016. 746906459439529985: Success
1017. 746872823977771008: Success
1018. 746818907684614144: Success
1019. 746790600704425984: Success
1020. 746757706116112384: Success
1021. 7467268980850360

1234. 712717840512598017: Success
1235. 712668654853337088: Success
1236. 712438159032893441: Success
1237. 712309440758808576: Success
1238. 712097430750289920: Success
1239. 712092745624633345: Success
1240. 712085617388212225: Success
1241. 712065007010385924: Success
1242. 711998809858043904: Success
1243. 711968124745228288: Success
1244. 711743778164514816: Success
1245. 711732680602345472: Success
1246. 711694788429553666: Success
1247. 711652651650457602: Success
1248. 711363825979756544: Success
1249. 711306686208872448: Success
1250. 711008018775851008: Success
1251. 710997087345876993: Success
1252. 710844581445812225: Success
1253. 710833117892898816: Success
1254. 710658690886586372: Success
1255. 710609963652087808: Success
1256. 710588934686908417: Success
1257. 710296729921429505: Success
1258. 710283270106132480: Success
1259. 710272297844797440: Success
1260. 710269109699739648: Success
1261. 710153181850935296: Success
1262. 710140971284037632: Success
1263. 71011701

1475. 693642232151285760: Success
1476. 693629975228977152: Success
1477. 693622659251335168: Success
1478. 693590843962331137: Success
1479. 693582294167244802: Success
1480. 693486665285931008: Success
1481. 693280720173801472: Success
1482. 693267061318012928: Success
1483. 693262851218264065: Success
1484. 693231807727280129: Success
1485. 693155686491000832: Success
1486. 693109034023534592: Success
1487. 693095443459342336: Success
1488. 692919143163629568: Success
1489. 692905862751522816: Success
1490. 692901601640583168: Success
1491. 692894228850999298: Success
1492. 692828166163931137: Success
1493. 692752401762250755: Success
1494. 692568918515392513: Success
1495. 692535307825213440: Success
1496. 692530551048294401: Success
1497. 692423280028966913: Success
1498. 692417313023332352: Success
1499. 692187005137076224: Success
1500. 692158366030913536: Success
1501. 692142790915014657: Success
1502. 692041934689402880: Success
1503. 692017291282812928: Success
1504. 69182033

1716. 680206703334408192: Success
1717. 680191257256136705: Success
1718. 680176173301628928: Success
1719. 680161097740095489: Success
1720. 680145970311643136: Success
1721. 680130881361686529: Success
1722. 680115823365742593: Success
1723. 680100725817409536: Success
1724. 680085611152338944: Success
1725. 680070545539371008: Success
1726. 680055455951884288: Success
1727. 679877062409191424: Success
1728. 679872969355714560: Success
1729. 679862121895714818: Success
1730. 679854723806179328: Success
1731. 679844490799091713: Success
1732. 679828447187857408: Success
1733. 679777920601223168: Success
1734. 679736210798047232: Success
1735. 679729593985699840: Success
1736. 679722016581222400: Success
1737. 679530280114372609: Success
1738. 679527802031484928: Success
1739. 679511351870550016: Success
1740. 679503373272485890: Success
1741. 679475951516934144: Success
1742. 679462823135686656: Success
1743. 679405845277462528: Success
1744. 679158373988876288: Success
1745. 67914876

1957. 673583129559498752: Success
1958. 673580926094458881: Success
1959. 673576835670777856: Success
1960. 673363615379013632: Success
1961. 673359818736984064: Success
1962. 673355879178194945: Success
1963. 673352124999274496: Success
1964. 673350198937153538: Success
1965. 673345638550134785: Success
1966. 673343217010679808: Success
1967. 673342308415348736: Success
1968. 673320132811366400: Success
1969. 673317986296586240: Success
1970. 673295268553605120: Success
1971. 673270968295534593: Success
1972. 673240798075449344: Success
1973. 673213039743795200: Success
1974. 673148804208660480: Success
1975. 672997845381865473: Success
1976. 672995267319328768: Success
1977. 672988786805112832: Success
1978. 672984142909456390: Success
1979. 672980819271634944: Success
1980. 672975131468300288: Success
1981. 672970152493887488: Success
1982. 672968025906282496: Success
1983. 672964561327235073: Success
1984. 672902681409806336: Success
1985. 672898206762672129: Success
1986. 67288442

2198. 668815180734689280: Success
2199. 668779399630725120: Success
2200. 668655139528511488: Success
2201. 668645506898350081: Success
2202. 668643542311546881: Success
2203. 668641109086707712: Success
2204. 668636665813057536: Success
2205. 668633411083464705: Success
2206. 668631377374486528: Success
2207. 668627278264475648: Success
2208. 668625577880875008: Success
2209. 668623201287675904: Success
2210. 668620235289837568: Success
2211. 668614819948453888: Success
2212. 668587383441514497: Success
2213. 668567822092664832: Success
2214. 668544745690562560: Success
2215. 668542336805281792: Success
2216. 668537837512433665: Success
2217. 668528771708952576: Success
2218. 668507509523615744: Success
2219. 668496999348633600: Success
2220. 668484198282485761: Success
2221. 668480044826800133: Success
2222. 668466899341221888: Success
2223. 668297328638447616: Success
2224. 668291999406125056: Success
2225. 668286279830867968: Success
2226. 668274247790391296: Success
2227. 66826890

In [8]:
len(success_tweets)

2345

Creating a dataframe from the json data 

In [9]:

start=time.time()

json_list=[]
tweet_id=[]
favorite_count=[]
retweet_count=[]

with open("tweet_json.txt") as file:
    json_txt= file.readlines()

for entry in json_txt:
    json_entry = json.loads(entry)
    json_list.append(json_entry)
    tweet_id.append(json_entry['id'])
    favorite_count.append(json_entry['favorite_count'])
    retweet_count.append(json_entry['retweet_count'])
    
end=time.time()
print(str(end-start)+ "seconds to complete")

0.871535062789917seconds to complete


In [10]:
df_3=pd.DataFrame({'tweet_id':tweet_id,
                "favorite_count": favorite_count,
                'retweet_count': retweet_count})

## Assess

### df_1 : twitter-archive-enhanced.csv

In [11]:
df_1.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [12]:
df_1.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

**Issues:**
- source is in html format and not text
- retweets and replies columns have a lot of Nan
- attributes like doggo, pupper, floofer, puppo need to be set as dummies

In [13]:
df_1.shape

(2356, 17)

In [14]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

**Issues:**
* Missing values
    - expanded_urls 
    - in_reply_to_status_id & in_reply_to_user_id
    - retweeted_status_id, retweeted_status_user_id & retweeted_status_timestamp 

* dtypes
    - tweet_id: int64
    - in_reply_to_status_id: float64
    - in_reply_to_user_id: float64
    - retweeted_status_id: float64
    - retweeted_status_user_id: float64


In [15]:
df_1.timestamp[666], type(df_1.timestamp[0])

('2016-10-24 15:53:19 +0000', str)

In [16]:
type((df_1.retweeted_status_timestamp[df_1.retweeted_status_timestamp.notna()].tolist())[0])

str

**Issue:**
- timestamp is a string
- retweeted_status_timestamp is a string

In [17]:
df_1.expanded_urls.sample(10)

259     https://twitter.com/dog_rates/status/843235543...
966     https://twitter.com/dog_rates/status/750383411...
1509    https://twitter.com/dog_rates/status/691459709...
1543    https://twitter.com/dog_rates/status/689557536...
2110    https://twitter.com/dog_rates/status/670444955...
1252    https://twitter.com/dog_rates/status/710844581...
882     https://twitter.com/dog_rates/status/760290219...
1998    https://twitter.com/dog_rates/status/672538107...
959     https://twitter.com/dog_rates/status/751251247...
584     https://twitter.com/dog_rates/status/800141422...
Name: expanded_urls, dtype: object

Issues:
- expanded_urls have multiple urls for some entries
- expanded_urls have whitespace for some entries


In [18]:
pd.set_option('display.max_colwidth', -1)
df_1.text.sample(10)

345     This is Emanuel. He's a h*ckin rare doggo. Dwells in a semi-urban environment. Round features make him extra collectible. 12/10 would so pet https://t.co/k9bzgyVdUT
1452    If you are aware of who is making these please let me know. 13/10 vroom vroom https://t.co/U0D1sbIDrG                                                               
350     This is Dutch. He dressed up as his favorite emoji for Valentine's Day. I've got heart eyes for his heart eyes. 13/10 https://t.co/BCbmFYLrse                       
1109    This is Terry. The harder you hug him the farther his tongue sticks out. 10/10 magical af https://t.co/RFToQQI8fJ                                                   
519     This is Phil. He's a father. A very good father too. 13/10 everybody loves Phil https://t.co/9p6ECXJMMu                                                             
1582    This is Baxter. He looks like a fun dog. Prefers action shots. 11/10 the last one is impeccable https://t.co/LHcH1yhhIb        

In [19]:
pd.reset_option('display.max_colwidth')

**Issues:**
- Some of the tweet texts have shortened urls included

In [20]:
df_1.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [21]:
df_1.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [22]:
df_1.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

Issues:
- rating_numerator has min value as 0 and high maximum value
    - checking value_counts reveals a good picture of range and outliers
- rating_denominator has min value as 0 and high maximum value
    - checking value_counts reveals less issues with denominator but we can still see some outliers

In [23]:
df_1.duplicated().sum(), df_1.tweet_id.duplicated().sum(), df_1.text.duplicated().sum()

(0, 0, 0)

In [24]:
df_1.timestamp.duplicated().sum(), df_1.expanded_urls.duplicated().sum()

(0, 137)

### df_2 : image-predictions.tsv

In [25]:
df_2.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

**Issues:**
- inconsistent alphabetic casing in p1, p2, p3
- p1, p2, p3 contain "_" instead of space
- non descriptive column names
- the table seems to be secondary ie supplement to the primary data assessed as df_1
- breed of dog needs to be concluded 

In [26]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


**Issues:**

- tweet_id is of int64 dtype

In [27]:
df_2.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [28]:
df_2.duplicated().sum()

0

### df_3: twitter API data

In [29]:
df_3.head()

favorite_count  retweet_count            tweet_id
0           38689           8558  892420643555336193
1           33164           6292  892177421306343426
2           24966           4174  891815181378084864
3           42074           8689  891689557279858688
4           40227           9451  891327558926688256

In [30]:
df_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2345 entries, 0 to 2344
Data columns (total 3 columns):
favorite_count    2345 non-null int64
retweet_count     2345 non-null int64
tweet_id          2345 non-null int64
dtypes: int64(3)
memory usage: 55.0 KB


In [31]:
df_3.describe()

favorite_count  retweet_count      tweet_id
count     2345.000000    2345.000000  2.345000e+03
mean      8044.695949    3015.050746  7.422940e+17
std      12105.024041    5015.248880  6.833642e+16
min          0.000000       0.000000  6.660209e+17
25%       1402.000000     605.000000  6.783802e+17
50%       3529.000000    1404.000000  7.189392e+17
75%       9947.000000    3510.000000  7.986979e+17
max     142995.000000   77131.000000  8.924206e+17

In [32]:
df_3.favorite_count.value_counts()

0        170
1233       5
456        4
694        3
526        3
258        3
2917       3
547        3
1472       3
2601       3
1404       3
1007       3
1160       2
3160       2
334        2
4747       2
1149       2
10255      2
5771       2
3232       2
4450       2
3363       2
2303       2
2314       2
1647       2
4362       2
1596       2
654        2
752        2
2262       2
        ... 
2678       1
629        1
149        1
567        1
2618       1
2620       1
573        1
27202      1
8771       1
1498       1
585        1
6732       1
4685       1
2638       1
25167      1
4689       1
14964      1
595        1
25173      1
2650       1
605        1
6752       1
609        1
29287      1
17001      1
5514       1
476        1
2670       1
8819       1
12287      1
Name: favorite_count, Length: 2025, dtype: int64

In [33]:
df_3.retweet_count.value_counts()

1301     5
373      5
1131     4
270      4
92       4
822      4
551      4
45       4
496      4
2611     4
390      4
316      4
588      4
1088     4
2880     4
528      4
131      4
807      3
93       3
3756     3
142      3
328      3
558      3
130      3
569      3
765      3
544      3
812      3
557      3
1495     3
        ..
2488     1
443      1
445      1
6590     1
4543     1
453      1
2506     1
2452     1
4489     1
2392     1
367      1
347      1
2404     1
357      1
570      1
361      1
2410     1
363      1
2412     1
2145     1
2438     1
6514     1
371      1
2420     1
375      1
383      1
39296    1
2434     1
389      1
0        1
Name: retweet_count, Length: 1718, dtype: int64

Issues:
- Order of columns is not as desired, tweet_id should be first
- favorite_count has 170 columns with value 0 which seems raises questions when followers are close to 7million and tweets are about cute dogs 

### Summarizing quality and tidiness issues

#### Quality Issues

##### df_1: twitter_archive 

1. dtypes
    - tweet_id: int64
    - in_reply_to_status_id: float64
    - in_reply_to_user_id: float64
    - retweeted_status_id: float64
    - retweeted_status_user_id: float64 

2. timestamps  
    - timestamps column is a string
    - retweeted_status_timestamp is a string
        - not an issue to worry if we drop retweets

3. source is in html format and not text

4. Missing values
    - in_reply_to_status_id & in_reply_to_user_id
    - retweeted_status_id, retweeted_status_user_id & retweeted_status_timestamp 
    - will create separate dataframes for these

5. expanded_urls 
    - have multiple urls for some entries
        - can be separated into different urls
        - gofundme urls can be extracted
    - expanded_urls have whitespace for some entries
    - missing entries for urls

6. Some of the tweet texts have shortened urls included
    - these are not needed thus they can be deleted from columns

7. rating_numerator
    - rating_numerator has min value as 0 and high maximum value
        - checking value_counts reveals a good picture of range and outliers
8. rating_denominator
   - rating_denominator has min value as 0 and high maximum value
        - checking value_counts reveals less issues with denominator but we can still see some outliers

##### df_2 : image-predictions.tsv
9. non descriptive column names
10. p1, p2, p3
    - inconsistent alphabetic casing 
    - contain "_" instead of space
11. tweet_id is of int64 dtype
12. breed of dog needs to be concluded 

##### df_3: twitter API data
There are only few issues with this since it is from a semi structured source. These issues can be ignored as this dataframe will be merged eventually
   - Order of columns is not as desired, tweet_id should be first
   - favorite_count has 170 columns with value 0 which seems raises questions when followers are close to 7 million and tweets are about cute dogs 


#### Tidiness Issues

1.	attributes like doggo, pupper, floofer, puppo seem out of place
2.	unwanted columns
3.	all three datasets need to be merged
4.	column order needs to be changed
5.  separate tables for retweets and replies
    - would have already done this in earlier steps

## Cleaning

In [34]:
# create a copy for each dataframe for cleaning
df_1c=df_1.copy()
df_2c=df_2.copy()
df_3c=df_3.copy()

### Issue 1: Fixing dtypes
- tweet_id: int64

Not fixing following since they will be dropped eventually due to presence of NaN
- in_reply_to_status_id: float64
- in_reply_to_user_id: float64
- retweeted_status_id: float64
- retweeted_status_user_id: float64

#### Define
Use astype to convert them into strings 

#### Code

In [35]:
df_1c['tweet_id'] = df_1c.tweet_id.astype(str)

#### Check

In [36]:
df_1c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null object
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(2), o

In [37]:
type(df_1c.tweet_id[0])

str

### Issue 2: timestamp and retweeted_status_timestamp of type str and not timestamp

#### Define
use pd.to_datetime to comvert to a datetime64 (timestamp type for scalar value)

#### Code

In [38]:
df_1c['timestamp'] = pd.to_datetime(df_1c.timestamp)
df_1c['retweeted_status_timestamp'] = pd.to_datetime(df_1c.retweeted_status_timestamp)


#### Check

In [39]:
df_1c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null object
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null datetime64[ns]
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null datetime64[ns]
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: datetim

### Issue 3: source is in html format and not text

#### Define
Use the split function and extract the source name and then replace with shortened name if needed

#### Code

In [40]:
df_1c['source']=df_1c.source.str.split('>', expand=True)[1]
df_1c['source']=df_1c.source.str.split('<', expand=True)[0]
df_1c.source.replace("Vine - Make a Scene", "Vine", inplace=True)

#### Check

In [41]:
df_1c.source.value_counts()

Twitter for iPhone    2221
Vine                    91
Twitter Web Client      33
TweetDeck               11
Name: source, dtype: int64

### Issue 4: retweets and replies columns have a lot of Nan

#### Define
drop retweets and make a separate table just in case we need it later
- use notna to create the new retweet dataframe 
- use isnull to mask and recreate the dataframe
- drop columns for retweets
- use notna to create the new replies dataframe 
- drop columns for retweets



#### Code

In [42]:
df_4 = df_1c[df_1c.retweeted_status_id.notna()]
df_5 = df_1c[df_1c.in_reply_to_status_id.notna()]
df_1c['is_reply']=df_1c.apply(lambda x: 1 if not np.isnan(x['in_reply_to_status_id']) else 0, axis=1)
df_1c = df_1c[df_1c.retweeted_status_id.isnull()]
df_1c.drop(columns=['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], inplace=True)

#### Check

In [43]:
df_4.shape, df_1c.shape

((181, 17), (2175, 13))

In [44]:
df_4.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19  888202515573088257                    NaN                  NaN   
32  886054160059072513                    NaN                  NaN   
36  885311592912609280                    NaN                  NaN   
68  879130579576475649                    NaN                  NaN   
73  878404777348136964                    NaN                  NaN   

             timestamp              source  \
19 2017-07-21 01:02:36  Twitter for iPhone   
32 2017-07-15 02:45:48  Twitter for iPhone   
36 2017-07-13 01:35:06  Twitter for iPhone   
68 2017-06-26 00:13:58  Twitter for iPhone   
73 2017-06-24 00:09:53  Twitter for iPhone   

                                                 text  retweeted_status_id  \
19  RT @dog_rates: This is Canela. She attempted s...         8.874740e+17   
32  RT @Athletics: 12/10 #BATP https://t.co/WxwJmv...         8.860537e+17   
36  RT @dog_rates: This is Lilly. She just paralle...         8.305833e+17   
68  RT @dog_rates: This is Emmy. She was adopted t...         8.780576e+17   
73  RT @dog_rates: Meet Shadow. In an attempt to r...         8.782815e+17   

    retweeted_status_user_id retweeted_status_timestamp  \
19              4.196984e+09        2017-07-19 00:47:34   
32              1.960740e+07        2017-07-15 02:44:07   
36              4.196984e+09        2017-02-12 01:04:29   
68              4.196984e+09        2017-06-23 01:10:23   
73              4.196984e+09        2017-06-23 16:00:04   

                                        expanded_urls  rating_numerator  \
19  https://twitter.com/dog_rates/status/887473957...                13   
32  https://twitter.com/dog_rates/status/886053434...                12   
36  https://twitter.com/dog_rates/status/830583320...                13   
68  https://twitter.com/dog_rates/status/878057613...                14   
73  https://www.gofundme.com/3yd6y1c,https://twitt...                13   

    rating_denominator    name doggo floofer pupper puppo  
19                  10  Canela  None    None   None  None  
32                  10    None  None    None   None  None  
36                  10   Lilly  None    None   None  None  
68                  10    Emmy  None    None   None  None  
73                  10  Shadow  None    None   None  None

In [45]:
df_5.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30   886267009285017600           8.862664e+17         2.281182e+09   
55   881633300179243008           8.816070e+17         4.738443e+07   
64   879674319642796034           8.795538e+17         3.105441e+09   
113  870726314365509632           8.707262e+17         1.648776e+07   
148  863427515083354112           8.634256e+17         7.759620e+07   

              timestamp              source  \
30  2017-07-15 16:51:35  Twitter for iPhone   
55  2017-07-02 21:58:53  Twitter for iPhone   
64  2017-06-27 12:14:36  Twitter for iPhone   
113 2017-06-02 19:38:25  Twitter for iPhone   
148 2017-05-13 16:15:35  Twitter for iPhone   

                                                  text  retweeted_status_id  \
30   @NonWhiteHat @MayhewMayhem omg hello tanner yo...                  NaN   
55   @roushfenway These are good dogs but 17/10 is ...                  NaN   
64                     @RealKentMurphy 14/10 confirmed                  NaN   
113  @ComplicitOwl @ShopWeRateDogs &gt;10/10 is res...                  NaN   
148  @Jack_Septic_Eye I'd need a few more pics to p...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp expanded_urls  \
30                        NaN                        NaT           NaN   
55                        NaN                        NaT           NaN   
64                        NaN                        NaT           NaN   
113                       NaN                        NaT           NaN   
148                       NaN                        NaT           NaN   

     rating_numerator  rating_denominator  name doggo floofer pupper puppo  
30                 12                  10  None  None    None   None  None  
55                 17                  10  None  None    None   None  None  
64                 14                  10  None  None    None   None  None  
113                10                  10  None  None    None   None  None  
148                12                  10  None  None    None   None  None

In [46]:
df_1c.head()

tweet_id           timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  is_reply  
0                  10   Phineas  None    None   None  None         0  
1                  10     Tilly  None    None   None  None         0  
2                  10    Archie  None    None   None  None         0  
3                  10     Darla  None    None   None  None         0  
4                  10  Franklin  None    None   None  None         0

In [47]:
df_1c.isnull().sum()

tweet_id               0
timestamp              0
source                 0
text                   0
expanded_urls         58
rating_numerator       0
rating_denominator     0
name                   0
doggo                  0
floofer                0
pupper                 0
puppo                  0
is_reply               0
dtype: int64

### Issue 5: expanded_urls have whitespaces and multiple urls for some entries, some values are NaN

#### Define 
- create a new column for url using tweet_id
- drop expanded urls

#### Code

In [48]:
df_1c['url']='https://twitter.com/dog_rates/status/' + df_1c['tweet_id']
df_1c.drop('expanded_urls', axis=1,inplace=True)

#### Check

In [49]:
df_1c.url.isnull().sum(), df_1c.url

(0, 0       https://twitter.com/dog_rates/status/892420643...
 1       https://twitter.com/dog_rates/status/892177421...
 2       https://twitter.com/dog_rates/status/891815181...
 3       https://twitter.com/dog_rates/status/891689557...
 4       https://twitter.com/dog_rates/status/891327558...
 5       https://twitter.com/dog_rates/status/891087950...
 6       https://twitter.com/dog_rates/status/890971913...
 7       https://twitter.com/dog_rates/status/890729181...
 8       https://twitter.com/dog_rates/status/890609185...
 9       https://twitter.com/dog_rates/status/890240255...
 10      https://twitter.com/dog_rates/status/890006608...
 11      https://twitter.com/dog_rates/status/889880896...
 12      https://twitter.com/dog_rates/status/889665388...
 13      https://twitter.com/dog_rates/status/889638837...
 14      https://twitter.com/dog_rates/status/889531135...
 15      https://twitter.com/dog_rates/status/889278841...
 16      https://twitter.com/dog_rates/status/8889172

In [50]:
df_1c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 13 columns):
tweet_id              2175 non-null object
timestamp             2175 non-null datetime64[ns]
source                2175 non-null object
text                  2175 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
is_reply              2175 non-null int64
url                   2175 non-null object
dtypes: datetime64[ns](1), int64(3), object(9)
memory usage: 237.9+ KB


### Issue 6: Some of the tweet texts have shortened urls included


#### Define
use replace and strip to get the desired result

#### Code

In [51]:
df_1c['text'] = df_1c.text.str.replace('https://.*', "", ).str.strip()

#### Check

In [52]:
df_1c.text.str.contains('https').sum()

0

In [53]:
df_1c.text[15]

"This is Oliver. You're witnessing one of his many brutal attacks. Seems to be playing with his victim. 13/10 fr*ckin frightening #BarkWeek"

### Issue 7&8: rating numerator and denominator have invalid ratings and unimportant outliers

#### Define
on post https://twitter.com/dog_rates/status/838085839343206401 WeRateDogs says they are looking for first 15/10 so we know that no rates are above 14, except for a few like Snoop Dogg or Satan's dog which have recieved 420/10 or 666/10. 

Since these are all outliers we can drop them

Some others are high numerators with high denominators, usually   litter of puppers or pack of doggos, these can be figured out by looking at denominator which we saw earlier didn't have too many outliers

- check all with denomiator neither 10 nor divisible by 10
    - note for change or drop 
- create a divisor column that will divide numerator and denominator for litters and packs
- check outliers and drop them

In [54]:
df_1d=df_1c.copy()

In [55]:
df_1c['rate']=df_1c.text.str.extract("(\d+\.?\d*/\d+)", expand=True)

In [56]:
df_1c['rate'].isnull().sum()

0

In [57]:
rate = df_1c['rate'].str.split('/')
df_1c['rating_numerator']= df_1c['rate'].str.split('/', expand=True)[0].astype(float)
df_1c['rating_denominator'] = df_1c['rate'].str.split('/', expand=True)[1].astype(float)
df_1c.drop('rate', axis=1)

tweet_id           timestamp              source  \
0     892420643555336193 2017-08-01 16:23:56  Twitter for iPhone   
1     892177421306343426 2017-08-01 00:17:27  Twitter for iPhone   
2     891815181378084864 2017-07-31 00:18:03  Twitter for iPhone   
3     891689557279858688 2017-07-30 15:58:51  Twitter for iPhone   
4     891327558926688256 2017-07-29 16:00:24  Twitter for iPhone   
5     891087950875897856 2017-07-29 00:08:17  Twitter for iPhone   
6     890971913173991426 2017-07-28 16:27:12  Twitter for iPhone   
7     890729181411237888 2017-07-28 00:22:40  Twitter for iPhone   
8     890609185150312448 2017-07-27 16:25:51  Twitter for iPhone   
9     890240255349198849 2017-07-26 15:59:51  Twitter for iPhone   
10    890006608113172480 2017-07-26 00:31:25  Twitter for iPhone   
11    889880896479866881 2017-07-25 16:11:53  Twitter for iPhone   
12    889665388333682689 2017-07-25 01:55:32  Twitter for iPhone   
13    889638837579907072 2017-07-25 00:10:02  Twitter for iPhone   
14    889531135344209921 2017-07-24 17:02:04  Twitter for iPhone   
15    889278841981685760 2017-07-24 00:19:32  Twitter for iPhone   
16    888917238123831296 2017-07-23 00:22:39  Twitter for iPhone   
17    888804989199671297 2017-07-22 16:56:37  Twitter for iPhone   
18    888554962724278272 2017-07-22 00:23:06  Twitter for iPhone   
20    888078434458587136 2017-07-20 16:49:33  Twitter for iPhone   
21    887705289381826560 2017-07-19 16:06:48  Twitter for iPhone   
22    887517139158093824 2017-07-19 03:39:09  Twitter for iPhone   
23    887473957103951883 2017-07-19 00:47:34  Twitter for iPhone   
24    887343217045368832 2017-07-18 16:08:03  Twitter for iPhone   
25    887101392804085760 2017-07-18 00:07:08  Twitter for iPhone   
26    886983233522544640 2017-07-17 16:17:36  Twitter for iPhone   
27    886736880519319552 2017-07-16 23:58:41  Twitter for iPhone   
28    886680336477933568 2017-07-16 20:14:00  Twitter for iPhone   
29    886366144734445568 2017-07-15 23:25:31  Twitter for iPhone   
30    886267009285017600 2017-07-15 16:51:35  Twitter for iPhone   
...                  ...                 ...                 ...   
2326  666411507551481857 2015-11-17 00:24:19  Twitter for iPhone   
2327  666407126856765440 2015-11-17 00:06:54  Twitter for iPhone   
2328  666396247373291520 2015-11-16 23:23:41  Twitter for iPhone   
2329  666373753744588802 2015-11-16 21:54:18  Twitter for iPhone   
2330  666362758909284353 2015-11-16 21:10:36  Twitter for iPhone   
2331  666353288456101888 2015-11-16 20:32:58  Twitter for iPhone   
2332  666345417576210432 2015-11-16 20:01:42  Twitter for iPhone   
2333  666337882303524864 2015-11-16 19:31:45  Twitter for iPhone   
2334  666293911632134144 2015-11-16 16:37:02  Twitter for iPhone   
2335  666287406224695296 2015-11-16 16:11:11  Twitter for iPhone   
2336  666273097616637952 2015-11-16 15:14:19  Twitter for iPhone   
2337  666268910803644416 2015-11-16 14:57:41  Twitter for iPhone   
2338  666104133288665088 2015-11-16 04:02:55  Twitter for iPhone   
2339  666102155909144576 2015-11-16 03:55:04  Twitter for iPhone   
2340  666099513787052032 2015-11-16 03:44:34  Twitter for iPhone   
2341  666094000022159362 2015-11-16 03:22:39  Twitter for iPhone   
2342  666082916733198337 2015-11-16 02:38:37  Twitter for iPhone   
2343  666073100786774016 2015-11-16 01:59:36  Twitter for iPhone   
2344  666071193221509120 2015-11-16 01:52:02  Twitter for iPhone   
2345  666063827256086533 2015-11-16 01:22:45  Twitter for iPhone   
2346  666058600524156928 2015-11-16 01:01:59  Twitter for iPhone   
2347  666057090499244032 2015-11-16 00:55:59  Twitter for iPhone   
2348  666055525042405380 2015-11-16 00:49:46  Twitter for iPhone   
2349  666051853826850816 2015-11-16 00:35:11  Twitter for iPhone   
2350  666050758794694657 2015-11-16 00:30:50  Twitter for iPhone   
2351  666049248165822465 2015-11-16 00:24:50  Twitter for iPhone   
2352  666044226329800704 2015-11-16 00:04:52  Twitter for iPhone   
2353

#### Reassess

In [58]:
df_1c[(df_1c.rating_numerator>100)]

tweet_id           timestamp              source  \
188   855862651834028034 2017-04-22 19:15:32  Twitter for iPhone   
189   855860136149123072 2017-04-22 19:05:32  Twitter for iPhone   
290   838150277551247360 2017-03-04 22:12:52  Twitter for iPhone   
313   835246439529840640 2017-02-24 21:54:03  Twitter for iPhone   
902   758467244762497024 2016-07-28 01:00:57  Twitter for iPhone   
979   749981277374128128 2016-07-04 15:00:45           TweetDeck   
1120  731156023742988288 2016-05-13 16:15:54  Twitter for iPhone   
1634  684225744407494656 2016-01-05 04:11:44  Twitter for iPhone   
1635  684222868335505415 2016-01-05 04:00:18  Twitter for iPhone   
1779  677716515794329600 2015-12-18 05:06:23  Twitter for iPhone   
2074  670842764863651840 2015-11-29 05:52:33  Twitter for iPhone   

                                                   text  rating_numerator  \
188   @dhmontgomery We also gave snoop dogg a 420/10...             420.0   
189   @s8n You tried very hard to portray this good ...             666.0   
290                                  @markhoppus 182/10             182.0   
313   @jonnysun @Lin_Manuel ok jomny I know you're e...             960.0   
902   Why does this never happen at my front door......             165.0   
979   This is Atticus. He's quite simply America af....            1776.0   
1120  Say hello to this unbelievably well behaved sq...             204.0   
1634  Two sneaky puppers were not initially seen, mo...             143.0   
1635  Someone help the girl is being mugged. Several...             121.0   
1779     IT'S PUPPERGEDDON. Total of 144/120 ...I think             144.0   
2074  After so many requests... here you go.\n\nGood...             420.0   

      rating_denominator     name doggo floofer pupper puppo  is_reply  \
188                 10.0     None  None    None   None  None         1   
189                 10.0     None  None    None   None  None         1   
290                 10.0     None  None    None   None  None         1   
313                  0.0     None  None    None   None  None         1   
902                150.0     None  None    None   None  None         0   
979                 10.0  Atticus  None    None   None  None         0   
1120               170.0     this  None    None   None  None         0   
1634               130.0     None  None    None   None  None         1   
1635               110.0     None  None    None   None  None         0   
1779               120.0     None  None    None   None  None         0   
2074                10.0     None  None    None   None  None         0   

                                                    url     rate  
188   https://twitter.com/dog_rates/status/855862651...   420/10  
189   https://twitter.com/dog_rates/status/855860136...   666/10  
290   https://twitter.com/dog_rates/status/838150277...   182/10  
313   https://twitter.com/dog_rates/status/835246439...   960/00  
902   https://twitter.com/dog_rates/status/758467244...  165/150  
979   https://twitter.com/dog_rates/status/749981277...  1776/10  
1120  https://twitter.com/dog_rates/status/731156023...  204/170  
1634  https://twitter.com/dog_rates/status/684225744...  143/130  
1635  https://twitter.com/dog_rates/status/684222868...  121/110  
1779  https://twitter.com/dog_rates/status/677716515...  144/120  
2074  https://twitter.com/dog_rates/status/670842764...   420/10

In [59]:
df_1c[((df_1c.rating_denominator%10)!=0)]

tweet_id           timestamp              source  \
342   832088576586297345 2017-02-16 04:45:50  Twitter for iPhone   
516   810984652412424192 2016-12-19 23:06:23  Twitter for iPhone   
1068  740373189193256964 2016-06-08 02:41:38  Twitter for iPhone   
1662  682962037429899265 2016-01-01 16:30:13  Twitter for iPhone   
1663  682808988178739200 2016-01-01 06:22:03  Twitter for iPhone   
2335  666287406224695296 2015-11-16 16:11:11  Twitter for iPhone   

                                                   text  rating_numerator  \
342            @docmisterio account started on 11/15/15              11.0   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...              24.0   
1068  After so many requests, this is Bretagne. She ...               9.0   
1662  This is Darrel. He just robbed a 7/11 and is i...               7.0   
1663  I'm aware that I could've said 20/16, but here...              20.0   
2335  This is an Albanian 3 1/2 legged  Episcopalian...               1.0   

      rating_denominator    name doggo floofer pupper puppo  is_reply  \
342                 15.0    None  None    None   None  None         1   
516                  7.0     Sam  None    None   None  None         0   
1068                11.0    None  None    None   None  None         0   
1662                11.0  Darrel  None    None   None  None         0   
1663                16.0    None  None    None   None  None         1   
2335                 2.0      an  None    None   None  None         0   

                                                    url   rate  
342   https://twitter.com/dog_rates/status/832088576...  11/15  
516   https://twitter.com/dog_rates/status/810984652...   24/7  
1068  https://twitter.com/dog_rates/status/740373189...   9/11  
1662  https://twitter.com/dog_rates/status/682962037...   7/11  
1663  https://twitter.com/dog_rates/status/682808988...  20/16  
2335  https://twitter.com/dog_rates/status/666287406...    1/2

- We can drop all tweets that have rating numerator above 100 and denominator as 10
- We can also drop tweets that have denominator not equal 10 or divisible by 10 
- Some entries seem to have an error which we will fix before dropping, they are documented below
    - 740373189193256964 Bretagne 14/10 (actually post says this is second such dog to get this, its last of 9/11 rescue dogs)
    - 682962037429899265 Darrel 10/10
    - 666287406224695296 Episcopalian 9/10

#### Code

In [60]:
A= {'740373189193256964': ['Bretagne', 14],
    '682962037429899265': ['Darrel',  10],
    '666287406224695296': ['Episcopalian', 9],
    '881633300179243008': [None, 13]}

In [61]:
for key,value in A.items():
    (df_1c.loc[df_1c['tweet_id'] == key, 'name' ])=value[0]
for key,value in A.items():
    (df_1c.loc[df_1c['tweet_id'] == key, 'rating_numerator' ])=value[1]
for key,value in A.items():
    (df_1c.loc[df_1c['tweet_id'] == key, 'rating_denominator' ])=10

In [62]:
drop_list=(df_1c[(df_1c.rating_numerator>100)& (df_1c.rating_denominator==10)].tweet_id).tolist()
drop_list2 = df_1c[((df_1c.rating_denominator%10)!=0)].tweet_id.tolist()
drop_list.extend(drop_list2)

In [63]:
drop_list

['855862651834028034',
 '855860136149123072',
 '838150277551247360',
 '749981277374128128',
 '670842764863651840',
 '832088576586297345',
 '810984652412424192',
 '682808988178739200']

In [64]:
for dropper in drop_list:
    df_1c.drop(df_1c[df_1c['tweet_id'] == dropper].index, inplace=True)

In [65]:
df_1c['divider'] = df_1d.rating_denominator/10

In [66]:
df_1c['rating_denominator']= df_1c['rating_denominator']/df_1c['divider']
df_1c['rating_numerator']= df_1c['rating_numerator']/df_1c['divider']
df_1c.drop('divider',axis=1, inplace=True)

In [67]:
df_1c.drop(df_1c.loc[(df_1c['rating_denominator']!=10)|(df_1c.rating_numerator<4)].index, inplace=True)

#### Check

In [68]:
df_1c.rating_numerator.value_counts()

12.00    504
10.00    444
11.00    431
13.00    308
9.00     156
8.00      98
7.00      53
14.00     43
5.00      34
6.00      32
4.00      15
11.27      1
15.00      1
9.50       1
13.50      1
11.26      1
9.75       1
Name: rating_numerator, dtype: int64

In [69]:
df_1d.rating_numerator.value_counts()

12      500
10      442
11      426
13      307
9       156
8        98
7        54
14       43
5        36
6        32
3        19
4        17
2         9
1         8
420       2
0         2
20        1
24        1
26        1
84        1
44        1
50        1
60        1
80        1
165       1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
15        1
17        1
27        1
45        1
75        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [70]:
df_1c.rating_denominator.value_counts(),

(10.0    2124
 Name: rating_denominator, dtype: int64,)

In [71]:
df_1c.shape

(2124, 14)

We can now drop the rating denominator column since all values are equal to 10

In [72]:
df_1c.drop('rating_denominator', axis=1, inplace=True)

### df_2 : image-predictions.tsv

In [73]:
df_2c.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

### Issue 9: renaming non descriptive columns

#### Define
Use pandas rename function to rename the columns

In [74]:
df_2c.rename(columns={'p1': 'p1_most_probable_breed',
              'p1_conf': 'p1_confidence',
              'p1_dog': 'p1_is_dog_breed',
              'p2': 'p2_most_probable_breed',
              'p2_conf': 'p2_confidence',
              'p2_dog': 'p2_is_dog_breed',
              'p3': 'p3_most_probable_breed',
              'p3_conf': 'p3_confidence',
              'p3_dog': 'p3_is_dog_breed'}, inplace=True)

#### Check

In [75]:
df_2c.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num  p1_most_probable_breed  p1_confidence  p1_is_dog_breed  \
0        1  Welsh_springer_spaniel       0.465074             True   
1        1                 redbone       0.506826             True   
2        1         German_shepherd       0.596461             True   
3        1     Rhodesian_ridgeback       0.408143             True   
4        1      miniature_pinscher       0.560311             True   

  p2_most_probable_breed  p2_confidence  p2_is_dog_breed  \
0                 collie       0.156665             True   
1     miniature_pinscher       0.074192             True   
2               malinois       0.138584             True   
3                redbone       0.360687             True   
4             Rottweiler       0.243682             True   

  p3_most_probable_breed  p3_confidence  p3_is_dog_breed  
0      Shetland_sheepdog       0.061428             True  
1    Rhodesian_ridgeback       0.072010             True  
2             bloodhound       0.116197             True  
3     miniature_pinscher       0.222752             True  
4               Doberman       0.154629             True

### Issue 10: p1, p2, p3 have inconsistent casing and "_" instead of spaces


#### Define
use str.lower() and str.replace function to change into the desired result

#### Code

In [76]:
for col_name in ['p1_most_probable_breed', 'p2_most_probable_breed', 'p3_most_probable_breed']:
    df_2c[col_name] = df_2c[col_name].str.replace('_', ' ')
    df_2c[col_name] = df_2c[col_name].str.title()

#### Check

In [77]:
df_2c.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num  p1_most_probable_breed  p1_confidence  p1_is_dog_breed  \
0        1  Welsh Springer Spaniel       0.465074             True   
1        1                 Redbone       0.506826             True   
2        1         German Shepherd       0.596461             True   
3        1     Rhodesian Ridgeback       0.408143             True   
4        1      Miniature Pinscher       0.560311             True   

  p2_most_probable_breed  p2_confidence  p2_is_dog_breed  \
0                 Collie       0.156665             True   
1     Miniature Pinscher       0.074192             True   
2               Malinois       0.138584             True   
3                Redbone       0.360687             True   
4             Rottweiler       0.243682             True   

  p3_most_probable_breed  p3_confidence  p3_is_dog_breed  
0      Shetland Sheepdog       0.061428             True  
1    Rhodesian Ridgeback       0.072010             True  
2             Bloodhound       0.116197             True  
3     Miniature Pinscher       0.222752             True  
4               Doberman       0.154629             True

### Issue 11: tweet_id is of int64 dtype

#### Define
use astype to change into str

#### Code

In [78]:
df_2c.tweet_id = df_2c.tweet_id.astype(str)

#### Check

In [79]:
type(df_2c.tweet_id[0])

str

In [80]:
df_2c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id                  2075 non-null object
jpg_url                   2075 non-null object
img_num                   2075 non-null int64
p1_most_probable_breed    2075 non-null object
p1_confidence             2075 non-null float64
p1_is_dog_breed           2075 non-null bool
p2_most_probable_breed    2075 non-null object
p2_confidence             2075 non-null float64
p2_is_dog_breed           2075 non-null bool
p3_most_probable_breed    2075 non-null object
p3_confidence             2075 non-null float64
p3_is_dog_breed           2075 non-null bool
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


### Issue 12: breed of dog needs to be concluded 

#### Define
create a new column breed_fin using apply and a defined function 

#### Code

In [81]:
def breed_define(row):
    if row['p1_is_dog_breed']:
        return (row['p1_most_probable_breed'])
    if row['p2_is_dog_breed']:
        return (row['p2_most_probable_breed'])
    if row['p3_is_dog_breed']:
        return (row['p3_most_probable_breed'])
    else:
        return np.nan

df_2c["breed_fin"]= df_2c.apply(breed_define, axis=1 )


#### Check

In [82]:
df_2c.isnull().sum()

tweet_id                    0
jpg_url                     0
img_num                     0
p1_most_probable_breed      0
p1_confidence               0
p1_is_dog_breed             0
p2_most_probable_breed      0
p2_confidence               0
p2_is_dog_breed             0
p3_most_probable_breed      0
p3_confidence               0
p3_is_dog_breed             0
breed_fin                 324
dtype: int64

## Tidying 

### Issue 1: doggo, puppo, pupper, floofer

#### Define
Create one column for stage and another for floofer (like a dummy column with 1 and 0). Interestingly only 10 were specified as floofer

#### Code

In [83]:
for col_name in ['pupper', 'puppo', 'doggo']:
    df_1c.loc[df_1c[col_name] == col_name, 'stage'] = col_name

df_1c.loc[df_1c.floofer == 'floofer', 'floofer'] = 1
df_1c.loc[df_1c.floofer != 1, 'floofer'] = 0

In [84]:
df_1c.stage.value_counts(), df_1c.floofer.value_counts()

(pupper    223
 doggo      87
 puppo      24
 Name: stage, dtype: int64, 0    2114
 1      10
 Name: floofer, dtype: int64)

In [85]:
df_1c.head()

tweet_id           timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24  Twitter for iPhone   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...              13.0   
1  This is Tilly. She's just checking pup on you....              13.0   
2  This is Archie. He is a rare Norwegian Pouncin...              12.0   
3  This is Darla. She commenced a snooze mid meal...              13.0   
4  This is Franklin. He would like you to stop ca...              12.0   

       name doggo  floofer pupper puppo  is_reply  \
0   Phineas  None        0   None  None         0   
1     Tilly  None        0   None  None         0   
2    Archie  None        0   None  None         0   
3     Darla  None        0   None  None         0   
4  Franklin  None        0   None  None         0   

                                                 url   rate stage  
0  https://twitter.com/dog_rates/status/892420643...  13/10   NaN  
1  https://twitter.com/dog_rates/status/892177421...  13/10   NaN  
2  https://twitter.com/dog_rates/status/891815181...  12/10   NaN  
3  https://twitter.com/dog_rates/status/891689557...  13/10   NaN  
4  https://twitter.com/dog_rates/status/891327558...  12/10   NaN

### Issue 2: Drop the unwanted columns

#### Define


#### Code

In [86]:
df_1c.drop(columns=['doggo', 'pupper', 'puppo', 'rate'], inplace=True)

#### Check

In [87]:
df_1c.head()

tweet_id           timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24  Twitter for iPhone   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...              13.0   
1  This is Tilly. She's just checking pup on you....              13.0   
2  This is Archie. He is a rare Norwegian Pouncin...              12.0   
3  This is Darla. She commenced a snooze mid meal...              13.0   
4  This is Franklin. He would like you to stop ca...              12.0   

       name  floofer  is_reply  \
0   Phineas        0         0   
1     Tilly        0         0   
2    Archie        0         0   
3     Darla        0         0   
4  Franklin        0         0   

                                                 url stage  
0  https://twitter.com/dog_rates/status/892420643...   NaN  
1  https://twitter.com/dog_rates/status/892177421...   NaN  
2  https://twitter.com/dog_rates/status/891815181...   NaN  
3  https://twitter.com/dog_rates/status/891689557...   NaN  
4  https://twitter.com/dog_rates/status/891327558...   NaN

### Issue 3: Merge all tables as needed

#### Define
use left merge 

#### Code

In [88]:
df=df_1c.merge(df_2c[['tweet_id', 'jpg_url', 'breed_fin' ]], how='left', on= "tweet_id")

#### Check

In [89]:
df

tweet_id           timestamp              source  \
0     892420643555336193 2017-08-01 16:23:56  Twitter for iPhone   
1     892177421306343426 2017-08-01 00:17:27  Twitter for iPhone   
2     891815181378084864 2017-07-31 00:18:03  Twitter for iPhone   
3     891689557279858688 2017-07-30 15:58:51  Twitter for iPhone   
4     891327558926688256 2017-07-29 16:00:24  Twitter for iPhone   
5     891087950875897856 2017-07-29 00:08:17  Twitter for iPhone   
6     890971913173991426 2017-07-28 16:27:12  Twitter for iPhone   
7     890729181411237888 2017-07-28 00:22:40  Twitter for iPhone   
8     890609185150312448 2017-07-27 16:25:51  Twitter for iPhone   
9     890240255349198849 2017-07-26 15:59:51  Twitter for iPhone   
10    890006608113172480 2017-07-26 00:31:25  Twitter for iPhone   
11    889880896479866881 2017-07-25 16:11:53  Twitter for iPhone   
12    889665388333682689 2017-07-25 01:55:32  Twitter for iPhone   
13    889638837579907072 2017-07-25 00:10:02  Twitter for iPhone   
14    889531135344209921 2017-07-24 17:02:04  Twitter for iPhone   
15    889278841981685760 2017-07-24 00:19:32  Twitter for iPhone   
16    888917238123831296 2017-07-23 00:22:39  Twitter for iPhone   
17    888804989199671297 2017-07-22 16:56:37  Twitter for iPhone   
18    888554962724278272 2017-07-22 00:23:06  Twitter for iPhone   
19    888078434458587136 2017-07-20 16:49:33  Twitter for iPhone   
20    887705289381826560 2017-07-19 16:06:48  Twitter for iPhone   
21    887517139158093824 2017-07-19 03:39:09  Twitter for iPhone   
22    887473957103951883 2017-07-19 00:47:34  Twitter for iPhone   
23    887343217045368832 2017-07-18 16:08:03  Twitter for iPhone   
24    887101392804085760 2017-07-18 00:07:08  Twitter for iPhone   
25    886983233522544640 2017-07-17 16:17:36  Twitter for iPhone   
26    886736880519319552 2017-07-16 23:58:41  Twitter for iPhone   
27    886680336477933568 2017-07-16 20:14:00  Twitter for iPhone   
28    886366144734445568 2017-07-15 23:25:31  Twitter for iPhone   
29    886267009285017600 2017-07-15 16:51:35  Twitter for iPhone   
...                  ...                 ...                 ...   
2094  666435652385423360 2015-11-17 02:00:15  Twitter for iPhone   
2095  666430724426358785 2015-11-17 01:40:41  Twitter for iPhone   
2096  666428276349472768 2015-11-17 01:30:57  Twitter for iPhone   
2097  666421158376562688 2015-11-17 01:02:40  Twitter for iPhone   
2098  666418789513326592 2015-11-17 00:53:15  Twitter for iPhone   
2099  666407126856765440 2015-11-17 00:06:54  Twitter for iPhone   
2100  666396247373291520 2015-11-16 23:23:41  Twitter for iPhone   
2101  666373753744588802 2015-11-16 21:54:18  Twitter for iPhone   
2102  666362758909284353 2015-11-16 21:10:36  Twitter for iPhone   
2103  666353288456101888 2015-11-16 20:32:58  Twitter for iPhone   
2104  666345417576210432 2015-11-16 20:01:42  Twitter for iPhone   
2105  666337882303524864 2015-11-16 19:31:45  Twitter for iPhone   
2106  666273097616637952 2015-11-16 15:14:19  Twitter for iPhone   
2107  666268910803644416 2015-11-16 14:57:41  Twitter for iPhone   
2108  666102155909144576 2015-11-16 03:55:04  Twitter for iPhone   
2109  666099513787052032 2015-11-16 03:44:34  Twitter for iPhone   
2110  666094000022159362 2015-11-16 03:22:39  Twitter for iPhone   
2111  666082916733198337 2015-11-16 02:38:37  Twitter for iPhone   
2112  666073100786774016 2015-11-16 01:59:36  Twitter for iPhone   
2113  666071193221509120 2015-11-16 01:52:02  Twitter for iPhone   
2114  666063827256086533 2015-11-16 01:22:45  Twitter for iPhone   
2115  666058600524156928 2015-11-16 01:01:59  Twitter for iPhone   
2116  666057090499244032 2015-11-16 00:55:59  Twitter for iPhone   
2117  666055525042405380 2015-11-16 00:49:46  Twitter for iPhone   
2118  666050758794694657 2015-11-16 00:30:50  Twitter for iPhone   
2119  666049248165822465 2015-11-16 00:24:50  Twitter for iPhone   
2120  666044226329800704 2015-11-16 00:04:52  Twitter for iPhone   
2121

#### Code

In [90]:
df_3c.tweet_id = df_3c.tweet_id.astype(str)

In [91]:
df=df.merge(df_3c, how='left', on= "tweet_id")

#### Check

In [92]:
df.shape

(2124, 14)

In [93]:
df.head()

tweet_id           timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24  Twitter for iPhone   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...              13.0   
1  This is Tilly. She's just checking pup on you....              13.0   
2  This is Archie. He is a rare Norwegian Pouncin...              12.0   
3  This is Darla. She commenced a snooze mid meal...              13.0   
4  This is Franklin. He would like you to stop ca...              12.0   

       name  floofer  is_reply  \
0   Phineas        0         0   
1     Tilly        0         0   
2    Archie        0         0   
3     Darla        0         0   
4  Franklin        0         0   

                                                 url stage  \
0  https://twitter.com/dog_rates/status/892420643...   NaN   
1  https://twitter.com/dog_rates/status/892177421...   NaN   
2  https://twitter.com/dog_rates/status/891815181...   NaN   
3  https://twitter.com/dog_rates/status/891689557...   NaN   
4  https://twitter.com/dog_rates/status/891327558...   NaN   

                                           jpg_url           breed_fin  \
0  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg                 NaN   
1  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg           Chihuahua   
2  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg           Chihuahua   
3  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg  Labrador Retriever   
4  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg              Basset   

   favorite_count  retweet_count  
0           38689           8558  
1           33164           6292  
2           24966           4174  
3           42074           8689  
4           40227           9451

#### Issue 4: rearrange columns

#### Define
rearrange columns by subsetting the whole dataframe using the order desired for columns

#### Code

In [94]:
df= df[['tweet_id', 'timestamp', 'source','url', 'jpg_url', 'is_reply', 'text', 'name', 'rating_numerator', 'breed_fin', 'stage', 'floofer', 'favorite_count', 'retweet_count' ]]


#### Check

In [95]:
df

tweet_id           timestamp              source  \
0     892420643555336193 2017-08-01 16:23:56  Twitter for iPhone   
1     892177421306343426 2017-08-01 00:17:27  Twitter for iPhone   
2     891815181378084864 2017-07-31 00:18:03  Twitter for iPhone   
3     891689557279858688 2017-07-30 15:58:51  Twitter for iPhone   
4     891327558926688256 2017-07-29 16:00:24  Twitter for iPhone   
5     891087950875897856 2017-07-29 00:08:17  Twitter for iPhone   
6     890971913173991426 2017-07-28 16:27:12  Twitter for iPhone   
7     890729181411237888 2017-07-28 00:22:40  Twitter for iPhone   
8     890609185150312448 2017-07-27 16:25:51  Twitter for iPhone   
9     890240255349198849 2017-07-26 15:59:51  Twitter for iPhone   
10    890006608113172480 2017-07-26 00:31:25  Twitter for iPhone   
11    889880896479866881 2017-07-25 16:11:53  Twitter for iPhone   
12    889665388333682689 2017-07-25 01:55:32  Twitter for iPhone   
13    889638837579907072 2017-07-25 00:10:02  Twitter for iPhone   
14    889531135344209921 2017-07-24 17:02:04  Twitter for iPhone   
15    889278841981685760 2017-07-24 00:19:32  Twitter for iPhone   
16    888917238123831296 2017-07-23 00:22:39  Twitter for iPhone   
17    888804989199671297 2017-07-22 16:56:37  Twitter for iPhone   
18    888554962724278272 2017-07-22 00:23:06  Twitter for iPhone   
19    888078434458587136 2017-07-20 16:49:33  Twitter for iPhone   
20    887705289381826560 2017-07-19 16:06:48  Twitter for iPhone   
21    887517139158093824 2017-07-19 03:39:09  Twitter for iPhone   
22    887473957103951883 2017-07-19 00:47:34  Twitter for iPhone   
23    887343217045368832 2017-07-18 16:08:03  Twitter for iPhone   
24    887101392804085760 2017-07-18 00:07:08  Twitter for iPhone   
25    886983233522544640 2017-07-17 16:17:36  Twitter for iPhone   
26    886736880519319552 2017-07-16 23:58:41  Twitter for iPhone   
27    886680336477933568 2017-07-16 20:14:00  Twitter for iPhone   
28    886366144734445568 2017-07-15 23:25:31  Twitter for iPhone   
29    886267009285017600 2017-07-15 16:51:35  Twitter for iPhone   
...                  ...                 ...                 ...   
2094  666435652385423360 2015-11-17 02:00:15  Twitter for iPhone   
2095  666430724426358785 2015-11-17 01:40:41  Twitter for iPhone   
2096  666428276349472768 2015-11-17 01:30:57  Twitter for iPhone   
2097  666421158376562688 2015-11-17 01:02:40  Twitter for iPhone   
2098  666418789513326592 2015-11-17 00:53:15  Twitter for iPhone   
2099  666407126856765440 2015-11-17 00:06:54  Twitter for iPhone   
2100  666396247373291520 2015-11-16 23:23:41  Twitter for iPhone   
2101  666373753744588802 2015-11-16 21:54:18  Twitter for iPhone   
2102  666362758909284353 2015-11-16 21:10:36  Twitter for iPhone   
2103  666353288456101888 2015-11-16 20:32:58  Twitter for iPhone   
2104  666345417576210432 2015-11-16 20:01:42  Twitter for iPhone   
2105  666337882303524864 2015-11-16 19:31:45  Twitter for iPhone   
2106  666273097616637952 2015-11-16 15:14:19  Twitter for iPhone   
2107  666268910803644416 2015-11-16 14:57:41  Twitter for iPhone   
2108  666102155909144576 2015-11-16 03:55:04  Twitter for iPhone   
2109  666099513787052032 2015-11-16 03:44:34  Twitter for iPhone   
2110  666094000022159362 2015-11-16 03:22:39  Twitter for iPhone   
2111  666082916733198337 2015-11-16 02:38:37  Twitter for iPhone   
2112  666073100786774016 2015-11-16 01:59:36  Twitter for iPhone   
2113  666071193221509120 2015-11-16 01:52:02  Twitter for iPhone   
2114  666063827256086533 2015-11-16 01:22:45  Twitter for iPhone   
2115  666058600524156928 2015-11-16 01:01:59  Twitter for iPhone   
2116  666057090499244032 2015-11-16 00:55:59  Twitter for iPhone   
2117  666055525042405380 2015-11-16 00:49:46  Twitter for iPhone   
2118  666050758794694657 2015-11-16 00:30:50  Twitter for iPhone   
2119  666049248165822465 2015-11-16 00:24:50  Twitter for iPhone   
2120  666044226329800704 2015-11-16 00:04:52  Twitter for iPhone   
2121

## Storing

We will be storing the following dataframes (names stored as mentioned after):

- df: twitter_archive_master.csv
- df_4: retweets_archive.csv
- df_5: replies_archive.csv

Amongst these twitter_archive_master.csv will be clean while retweets_archive.csv and replies_archive.csv will be in original format since they are not used for the analysis. 

In [96]:
df.to_csv('twitter_archive_master.csv', index=False)
df_4.to_csv('retweets_archive.csv', index=False)
df_5.to_csv('replies_archive.csv', index=False)
